File: psa-1-iskalna-drevesa/predavanja/iskalnodrevo.py
class Drevo:
    def __str__(self, zamik=""):
        if self.prazno:
            return zamik + "x"
        else:
            return "\n".join([
                self.desno.__str__(zamik=zamik+"   "),
                zamik + str(self.vrednost),
                self.levo.__str__(zamik=zamik+"   ")
            ])


class IskalnoDrevo(Drevo):
    def __init__(self, *elementi):
        self.prazno = True
        for x in elementi:
            self.dodaj(x)

    def dodaj(self, x):
        if self.prazno:
            self.prazno = False
            self.levo = IskalnoDrevo()
            self.vrednost = x
            self.desno = IskalnoDrevo()
        elif x < self.vrednost:
            self.levo.dodaj(x)
        elif x > self.vrednost:
            self.desno.dodaj(x)
        else:
            pass

    def poisci(self, x):
        if self.prazno:
            return False
        elif x < self.vrednost:
            return self.levo.poisci(x)
        elif x > self.vrednost:
            return self.desno.poisci(x)
        else:
            return True

    def izbrisi(self, x):
        if self.prazno:
            pass
        elif x < self.vrednost:
            self.levo.izbrisi(x)
        elif x > self.vrednost:
            self.desno.izbrisi(x)
        else:
            self.izbrisiKoren()

    def najbolj_levi(self):
        if self.levo.prazno:
            return self.vrednost
        else:
            return self.levo.najbolj_levi()

    def izbrisiKoren(self):
        if self.levo.prazno:
            self.zamenjaj(self.desno)
        elif self.desno.prazno:
            self.zamenjaj(self.levo)
        else:
            vrednost0 = self.desno.najbolj_levi()
            self.vrednost = vrednost0
            self.desno.izbrisi(vrednost0)

    def zamenjaj(self, drevo):
        if drevo.prazno:
            if self.prazno:
                pass
            else:
                del self.vrednost
                del self.levo
                del self.desno
                self.prazno = True
        else:
            self.vrednost = drevo.vrednost
            self.levo = drevo.levo
            self.desno = drevo.desno
            self.prazno = drevo.prazno

File: psa-1-iskalna-drevesa/predavanja/IskalnoDrevo.hs
module IskalnoDrevo where

    data Drevo a = Prazno
                 | Sestavljeno (Drevo a) a (Drevo a)

    instance Show a => Show (Drevo a) where
        show d = show' "" d
          where
            show' zamik Prazno = zamik ++ "x"
            show' zamik (Sestavljeno levo vrednost desno) =
                init $ unlines [
                    show' (zamik ++ "   ") desno,
                    zamik ++ show vrednost,
                    show' (zamik ++ "   ") levo
                ]

    iskalnoDrevo :: (Ord a) => [a] -> Drevo a
    iskalnoDrevo = foldl dodaj Prazno

    dodaj :: (Ord a) => Drevo a -> a -> Drevo a
    dodaj Prazno x = Sestavljeno Prazno x Prazno
    dodaj (Sestavljeno levo vrednost desno) x
        | x < vrednost = Sestavljeno (dodaj levo x) vrednost desno
        | x > vrednost = Sestavljeno levo vrednost (dodaj desno x)
        | otherwise = Sestavljeno levo vrednost desno

    poisci :: (Ord a) => Drevo a -> a -> Bool
    poisci Prazno _ = False
    poisci (Sestavljeno levo vrednost desno) x
       | x < vrednost = poisci levo x
       | x > vrednost = poisci desno x
       | otherwise = True

    izbrisi :: (Ord a) => Drevo a -> a -> Drevo a
    izbrisi Prazno _ = Prazno
    izbrisi (Sestavljeno levo vrednost desno) x
        | x < vrednost = Sestavljeno (izbrisi levo x) vrednost desno
        | x > vrednost = Sestavljeno levo vrednost (izbrisi desno x)
        | otherwise = izbrisiKoren (Sestavljeno levo vrednost desno)

    najboljLevi :: Drevo a -> a
    najboljLevi (Sestavljeno Prazno vrednost _) = vrednost
    najboljLevi (Sestavljeno levo _ _) = najboljLevi levo

    izbrisiKoren :: (Ord a) => Drevo a -> Drevo a
    izbrisiKoren (Sestavljeno Prazno vrednost desno) = desno
    izbrisiKoren (Sestavljeno levo vrednost Prazno) = levo
    izbrisiKoren (Sestavljeno levo vrednost desno) =
        let vrednost' = najboljLevi desno in
        Sestavljeno levo vrednost' (izbrisi desno vrednost')
